### PR TITLE
[TCFMM-14] Add property tests on range oracles

### PR DESCRIPTION
**Pull request by Martoon-00**
[//]: # (This is a template of a pull request template.)
[//]: # (You should modify it considering specifics of a particular repository and put it there.)
[//]: # (Comments like this are meta-comments, they shouldn't be present in the final template.)
[//]: # (Comments starting with '

[//]: # (Here you can add a link to the corresponding issue tracker, e. g. https://issues.serokell.io/issue/AD-)
[//]: # (For GitHub/GitLab issues it is better to use just hash symbol or exclamation mark as it is more resistant to repo movements)
[//]: # (In this case please also prefix the link with "Resolves" keyword)
[//]: # (See https://docs.gitlab.com/ee/user/project/issues/managing_issues.html#closing-issues-automatically)
[//]: # (See https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords)
## Related issue(s)

Resolves TCFMM-14 partially

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [Specification](../tree/master/docs/specification.md)
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

[//]: # (Update link to style guide if necesary or remove if it's not present)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

File: haskell/test/Test/Util.hs [hunk 1]
```diff
@@ -349,24 +351,12 @@ actual @~= expected = do
...
-    let X r = adjustScale @30 rx
-    let X a = adjustScale @30 ax
-    assert (l - 1 <= a && a <= r + 1) $
-      unlinesF
-        [ "Failed approximate in-range check"
-        , "━━ Range (rhs) ━━"
-        , build (lx, rx)
-        , "━━ Got value (lhs) ━━"
-        , build ax
-        ]
```
Comment (Martoon-00) [inline haskell/test/Test/Util.hs:369]: This turned out to be unnecessary, and a similar functionality is provided by some other functions now.

**Comment by Martoon-00** [comment]
Btw the test takes almost 40 seconds to complete... which is even greater than property tests on FA2 functionality.

File: haskell/test/Test/RangeOracles.hs [hunk 2]
```diff
@@ -108,4 +115,132 @@ test_ValuesSanity = testGroup "Values are sane"
...
+            -- Check cumulatives inside snapshot
+
+            let allTickRanges = do
+                  tick : otherTicks <- tails allTicks
+                  tick2 <- otherTicks
+                  return (tick, tick2)
+
+            s <- getFullStorage cfmm
+            let curTickIndex = sCurTickIndex s
+            let liquidityOfCoveringPositions = sum do
```
[the remark this answers is not in the document]
Comment (Martoon-00) [reply haskell/test/Test/RangeOracles.hs:218]: It should be, but who knows whether this property holds in the contract. Taking the oracle's purposes into account (for instance, they are to be used for liquidity mining), AFAIU oracles must report the time-weighted sum of liquidities of all active positions, and I followed this definition as the specification to test against.

[the remark this answers is not in the document]
Comment (dcastro) [reply haskell/test/Test/RangeOracles.hs:218]: Hm sounds like this should be a property in `checkStorageInvariants` :open_mouth: 

What do you think about moving this code there, and assert that it's equal to `sLiquidity s`? Then, you could simply use `sLiquidity s` here.

[the remark this answers is not in the document]
Comment (Martoon-00) [reply haskell/test/Test/RangeOracles.hs:218]: Sounds good!

File: haskell/test/Test/Invariants.hs [hunk 3]
```diff
@@ -107,13 +109,20 @@ checkStorageInvariants st = do
...
         & sum
-  sLiquidity st @== fromIntegral @Integer @Natural expectedLiquidity
+  sLiquidity st @== fromIntegral @Integer @Natural liquiditiyAfterPriorTicks
+
+  -- Invariant 2.2.
+  let liquidityOfCoveringPositions = sum do
+        PositionState{..} <- elems (bmMap $ sPositions st)
+        guard (curTickIndex `inTicksRange` (psLowerTickIndex, psUpperTickIndex))
+        return psLiquidity
+  sLiquidity st @== liquidityOfCoveringPositions
```
Comment (dcastro) [inline haskell/test/Test/Invariants.hs:125]: Nitpicking :stuck_out_tongue: 

```suggestion
  let liquidityOfActivePositions = sum do
        PositionState{..} <- elems (bmMap $ sPositions st)
        guard (curTickIndex `inTicksRange` (psLowerTickIndex, psUpperTickIndex))
        return psLiquidity
  sLiquidity st @== liquidityOfActivePositions
```

**Pull request merged**